lightning/src/util/persist.rs

```diff
@@ -0,0 +1,81 @@
+use bitcoin::hashes::hex::ToHex;
+use prelude::*;
+use io::{self};
+
+use crate::{chain::{keysinterface::{Sign, KeysInterface}, self, transaction::{OutPoint}, chaininterface::{BroadcasterInterface, FeeEstimator}, chainmonitor::{Persist, MonitorUpdateId}, channelmonitor::{ChannelMonitor, ChannelMonitorUpdate}}, ln::channelmanager::ChannelManager, routing::network_graph::NetworkGraph};
+
+use super::{logger::Logger, ser::Writeable};
+
+/// Trait that handles persisting a [`ChannelManager`] and [`NetworkGraph`] to disk.
+pub trait Persister<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> where
+M::Target: 'static + chain::Watch<Signer>,
+T::Target: 'static + BroadcasterInterface,
+K::Target: 'static + KeysInterface<Signer = Signer>,
+F::Target: 'static + FeeEstimator,
+L::Target: 'static + Logger
+{
+	/// Persist the given ['ChannelManager'] to disk, returning an error if persistence failed.
+	fn persist_manager(&self, channel_manager: &ChannelManager<Signer, M, T, K, F, L>) -> Result<(), io::Error>;
+
+	/// Persist the given [`NetworkGraph`] to disk, returning an error if persistence failed.
+	fn persist_graph(&self, network_graph: &NetworkGraph) -> Result<(), io::Error>;
+}
+
+/// Trait for a key-value store for persisting some writeable object at some key
+pub trait KVStorePersister {
+	/// Persist the given writeable using the provided key
+	fn persist<W: Writeable>(&self, key: String, object: &W) -> io::Result<()>;
+	/// Get the key for persisting a channel manager
+	fn get_channel_manager_key(&self) -> String { String::from("manager") }
+	/// Get the key for persisting a network graph
+	fn get_network_graph_key(&self) -> String { String::from("network_graph") }
+	/// Get the key for persisting a channel monitor from it's funding txo
+	fn get_channel_monitor_key(&self, funding_txo: &OutPoint) -> String {
+		format!("monitors/{}_{}", funding_txo.txid.to_hex(), funding_txo.index)
+	}
+}
+
+impl<A: KVStorePersister, Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> Persister<Signer, M, T, K, F, L> for A 
+where 
+M::Target: 'static + chain::Watch<Signer>,
+T::Target: 'static + BroadcasterInterface,
+K::Target: 'static + KeysInterface<Signer = Signer>,
+F::Target: 'static + FeeEstimator,
+L::Target: 'static + Logger
+{
+	/// Persist the given ['ChannelManager'] to disk, returning an error if persistence failed.
+	fn persist_manager(&self, channel_manager: &ChannelManager<Signer, M, T, K, F, L>) -> Result<(), io::Error> {
+		self.persist(self.get_channel_manager_key(), channel_manager)
+	}
+
+	/// Persist the given [`NetworkGraph`] to disk, returning an error if persistence failed.
+	fn persist_graph(&self, network_graph: &NetworkGraph) -> Result<(), io::Error> {
+		self.persist(self.get_network_graph_key(), network_graph)
+	}
+}
+
+impl<ChannelSigner: Sign, K: KVStorePersister> Persist<ChannelSigner> for K {
+	// TODO: We really need a way for the persister to inform the user that its time to crash/shut
```

Comment:
Lets open an issue for this? Using the `Event::ChannelClosed` event may make sense here, I think.

Reply:
https://github.com/lightningdevkit/rust-lightning/issues/1424